### PR TITLE
#474 - Automaticly Deploy to Pypi

### PR DESCRIPTION
**Pull request by robons**
- The main branch build now pushes to testpypi, e.g. [0.1.0.dev1090638346766807154676960913590186434387962801949](https://test.pypi.org/project/csvcubed/0.1.0.dev1090638346766807154676960913590186434387962801949/)
  - The long number after `dev` is the decimalised commit hash for HEAD of the branch. 
- The release tag builds push the whl to pypi automatically now  
  - The release process documentation needs to be updated so that windows testing happens before the tag is built in Jenkins from now on.

Partially satisties #474

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -118,6 +130,35 @@ pipeline {
...
+            when {
+                branch 'main'
+            }
+            steps {
+                script {
+                    sh "twine check dist/csvcubed*.whl"
+
+                    withCredentials([usernamePassword(credentialsId: 'testpypi-robons', usernameVariable:'TWINE_USERNAME', passwordVariable: 'TWINE_PASSWORD')]) {
+                        sh 'twine upload -r testpypi dist/csvcubed*.whl'
+                    }
```
Comment (GDonRanasinghe) [inline Jenkinsfile:143]: @robons This is self-explanatory from the name of `usernameVariable` and `passwordVariable`. But I would like to double-check. The `TWINE_USERNAME ` and `TWINE_PASSWORD` are env variables pointing to the actual username and password securely stored in jenkins right? And they are not the actual password and username?

File: Jenkinsfile [hunk 2]
```diff
@@ -118,6 +130,35 @@ pipeline {
...
+                buildingTag()
+                tag pattern: "v\\d+\\.\\d+\\.\\d+(-RC\\d)?", comparator: "REGEXP"
+            }
+            steps {
+                script {
+                    sh "twine check dist/csvcubed*.whl"
+
+                    withCredentials([usernamePassword(credentialsId: 'pypi-robons', usernameVariable:'TWINE_USERNAME', passwordVariable: 'TWINE_PASSWORD')]) {
+                        sh 'twine upload dist/csvcubed*.whl'
+                    }
```
Comment (GDonRanasinghe) [inline Jenkinsfile:158]: @robons This is self-explanatory from the name of usernameVariable and passwordVariable. But I would like to double-check. The TWINE_USERNAME and TWINE_PASSWORD are env variables pointing to the actual username and password securely stored in jenkins right? And they are not the actual password and username?

**Review by GDonRanasinghe — COMMENTED** [review]
@robons Left a minor comment on the PR around twine username and password. If you could answer my question there, I will approve - all the rest if all good. My question is more of a clarification for me.

File: Jenkinsfile (hunk 1 above)
Comment (robons) [reply Jenkinsfile:143]: https://twine.readthedocs.io/en/stable/index.html#environment-variables

>   TWINE_USERNAME - the username to use for authentication to the repository.
>    TWINE_PASSWORD - the password to use for authentication to the repository.

Yes, these variables are where we store the username and password that twine needs to authenticate with pypi.

**Pull request merged**